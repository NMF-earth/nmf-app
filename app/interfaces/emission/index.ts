import {
  Emission,
  EmissionEnum,
  EmissionPurchaseEnum
} from "./emission.interface";

export {
  Emission,
  EmissionEnum,
  EmissionPurchaseEnum
};
