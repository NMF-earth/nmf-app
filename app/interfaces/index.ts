import {
  Emission,
  EmissionEnum,
  EmissionPurchaseEnum
} from "./emission";

export {
  Emission,
  EmissionEnum,
  EmissionPurchaseEnum
};
