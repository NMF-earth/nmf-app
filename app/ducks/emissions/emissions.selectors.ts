import { filter, propEq, find, pathOr, pipe } from "ramda";
import { namespace } from "./emissions.slice";
import { EmissionEnum } from "../../interfaces";

const getAllEmissions = pathOr([], [namespace]);

/* TODO: add tests for this selector + check if working irl */
const getEmissionById = (state, id: number) =>
  find(propEq("id", id))(state[namespace]);

const getEmissionsToMitigate = state =>
  pipe(getAllEmissions, filter(propEq("isMitigated", false)))(state);

const getEmissionsMitigated = state =>
  pipe(getAllEmissions, filter(propEq("isMitigated", true)))(state);

const isTransportEmission = emission =>
  emission.emissionType === EmissionEnum.transport;

const isFoodEmission = emission => emission.emissionType === EmissionEnum.food;

const isCustomEmission = emission =>
  emission.emissionType === EmissionEnum.custom;

/* TODO: add tests for getTransportEmissions/getFoodEmissions/getCustomEmissions */
const getTransportEmissions = pipe(
  getAllEmissions,
  filter(isTransportEmission)
);

const getFoodEmissions = pipe(getAllEmissions, filter(isFoodEmission));

const getCustomEmissions = pipe(getAllEmissions, filter(isCustomEmission));

export default {
  getTransportEmissions,
  getFoodEmissions,
  getCustomEmissions,
  getEmissionById,
  getEmissionsToMitigate,
  getEmissionsMitigated
};
