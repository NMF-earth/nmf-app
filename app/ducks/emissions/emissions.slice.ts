import { createSlice, PayloadAction } from "@reduxjs/toolkit";

import { Emission } from "../../interfaces";

const initialState: Emission[] = [];

const emissions = createSlice({
  name: "emissions",
  initialState,
  reducers: {
    createEmission(state, action: PayloadAction<Emission>) {
      state.push(action.payload);
    },
    /* TODO: add test for deleteEmissionById and test if it's working irl */
    deleteEmissionById(state, action: PayloadAction<string>) {
      state = state.filter(item => item.id !== action.payload);
    }
  }
});

const { createEmission, deleteEmissionById } = emissions.actions;

export const actions = { createEmission, deleteEmissionById };

export const namespace = emissions.name;

export const reducer = emissions.reducer;
