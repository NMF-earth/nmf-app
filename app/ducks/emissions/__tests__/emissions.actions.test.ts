import emissions from "../";
import { Emission, EmissionEnum } from "../../../interfaces";
import { TransportEnum } from "carbon-footprint";

describe("Emissions actions should", () => {
  it("export expected actions", () =>
    expect(emissions.actions).toMatchSnapshot());

  it("be able to create an emission", () => {
    const emission: Emission = {
      id: "123",
      creationDate: "2020-01-26T11:04:55.334Z",
      emissionType: EmissionEnum.custom,
      emissionModelType: TransportEnum.boat,
      value: 200,
      isMitigated: false
    };

    const expectedAction = {
      type: emissions.actions.createEmission.toString(),
      payload: emission
    };

    expect(emissions.actions.createEmission(emission)).toEqual(expectedAction);
  });

  it("be able to delete an emission", () => {
    const expectedAction = {
      type: emissions.actions.deleteEmissionById.toString(),
      payload: "123"
    };

    expect(emissions.actions.deleteEmissionById("123")).toEqual(expectedAction);
  });
});
