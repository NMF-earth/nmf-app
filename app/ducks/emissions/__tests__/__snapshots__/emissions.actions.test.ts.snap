// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Emissions actions should export expected actions 1`] = `
{
  "createEmission": [Function],
  "deleteAllEmissions": [Function],
  "deleteEmission": [Function],
  "loadEmissionsFromImport": [Function],
  "toggleIsMitigated": [Function],
}
`;
